moped-database/migrations/1691605248601_council-district-trigger/up.sql

```diff
@@ -0,0 +1,150 @@
+-- add spatial index to this layer
+CREATE INDEX layer_council_district_geom_idx
+  ON layer_council_district
+  USING GIST (geography);
+
+-- create table to hold feature-district associations
+CREATE TABLE public.features_council_districts (
+    id serial PRIMARY KEY,
+    feature_id integer NOT NULL,
+    council_district_id integer NOT NULL,
+    FOREIGN KEY ("council_district_id") REFERENCES "public"."layer_council_district" ("council_district")
+        ON UPDATE CASCADE ON DELETE CASCADE
+);
+
+ALTER TABLE public.features_council_districts
+    ADD CONSTRAINT unique_feature_council_district UNIQUE ("feature_id", "council_district_id");
+
+-- create functions to mange point and line type associations
+CREATE OR REPLACE FUNCTION public.update_line_council_district()
+    RETURNS TRIGGER
+    LANGUAGE plpgsql
+    AS $$
+BEGIN
+    -- currently the moped editor never updates a geometry, it marks a feature as is_deleted
+    -- and inserts a new feature. but we should test if a geometry has changed to be safe.
+    IF TG_OP = 'UPDATE' AND ST_Equals(NEW.geography::geometry, OLD.geography::geometry) THEN
+        RETURN NEW;
+    END IF;
+    -- delete previous district associations
+    -- again, because the moped editor only inserts new geometries, there will typically
+    -- not be any old feature-district associations to delete
+    DELETE FROM features_council_districts WHERE feature_id = NEW.id;
+    -- insert new district associations
+    WITH inserts_todo AS (
+        SELECT
+            NEW.id AS feature_id,
+            districts.council_district AS council_district_id
+        FROM
+            layer_council_district AS districts
+        WHERE
+            ST_Intersects(districts.geography, NEW.geography)
+        OR
+            ST_Crosses(districts.geography::geometry, NEW.geography::geometry)
```

Comment:
together, the combined intersects and crosses methods cover all the scenarios for lines:
- Line is contained by the polygon 
- Line is exactly tracing the edge of a polygon
- Line passes through the polygon (but has no vertices inside it)

Reply:
This is tricky stuff! Thanks for the context and comments here. 🙏

Reply:
Also TIL that `TG_OP` is available to handle different operations in triggers. 🔖